themes/finna2/js/finna-model-viewer.js

```diff
@@ -1,529 +1,895 @@
+      properties: {
+        rotation: (object, pointers, value) => {
+          if (typeof pointers[2] !== 'undefined') {
+            switch (pointers[2]) {
+            case '_x':
+              object.rotation.x = THREE.Math.degToRad(value);
+              break;
+            case '_y':
+              object.rotation.y = THREE.Math.degToRad(value);
+              break;
+            case '_z':
+              object.rotation.z = THREE.Math.degToRad(value);
+              break;
+            }
           }
-          _.viewerStateInfo.html(loaded);
         }
       },
-      function onError(/*error*/) {
-        if (_.viewerStateInfo) {
-          _.viewerStateInfo.html('Error');
+      functions: {
+        lightMenuCreated: function lightMenuCreated(menu) {
+          if (this.menumode === 'basic') {
+            return;
+          }
+          const addLight = this.createButton('add-light', 'add-light', 'Add light');
+          menu.holder.append(addLight);
+          addLight.addEventListener('click', () => {
+            addLight.style.display = 'none';
+            this.menu.removeEventListener('change', this.updateFunction);
+            const templateClone = menu.template.cloneNode(true);
+            const div = this.createDiv('setting-child');
+            const span = document.createElement('span');
+            span.textContent = 'name';
+            div.append(span, this.createInput('text', 'light-name', ''));
+            const selectDiv = this.createDiv('setting-child');
+            const select = this.createSelect(this.options.lightTypeMappings, 'light-type', 'SpotLight');
+            const selectSpan = document.createElement('span');
+            selectSpan.textContent = this.translate('type');
+            selectDiv.append(selectSpan, select);
+            const saveLight = this.createButton('save-light', 'save-light', 'Save');
+            const form = templateClone.querySelector('form');
+            form.addEventListener('submit', (e) => e.preventDefault());
+            form.prepend(saveLight, selectDiv, div);
+            templateClone.classList.remove('template', 'hidden');
+            menu.holder.prepend(templateClone);
+            saveLight.addEventListener('click', e => {
+              const saveForm = e.target.closest('form');
+              if (saveForm) {
+                this.addLight(saveForm);
+                templateClone.parentNode.removeChild(templateClone);
+                addLight.style.display = null;
+                this.menu.addEventListener('change', this.updateFunction);
+              }
+            });
+          });
+        },
+        fileMenuCreated: function fileMenuCreated(menu) {
+          const exportButton = this.createButton('button export-model', 'export', this.translate('Export .glb'));
+          exportButton.addEventListener('click', () => {
+            this.menu.removeEventListener('change', this.updateFunction);
+            self.scene.children.forEach((child) => {
```

Comment:
this ja self, sekakäyttöä? Voisiko tässä metodissa käyttää vain this? Tuossa alempana on ainakin myös `            self.scene.add(cameraSave);`.